### PR TITLE
MechComp Componentization Refactor

### PR DESCRIPTION
**Pull request by MarkNais**
[CHORE][FEATURE]
## About the PR 
Refactors MechComp behaviour to be DM-Component based, along with some other changes.
Notable changes:
- mechanics_holder is now a component datum. 
  - It's disposing is improved.
  - It's almost setup well enough that the component could be transferred between objects. The use case for this is unclear, so it is not implemented.
  - Increased code reuse through more function calls.
  - Devices can now only have one connection to to each other device: I.E. You cannot connect a single button to each input of a selection component, only a single input.
  - Made some documentation at the top of `mechComp_signals.dm` for easier pick-up. 
- Changed a bunch of `if(signals.Find(input.signal))` to `if(input.signal in signals)` and similar lines.
- Device cooldowns were changed from `SPAWN_DBG()` to `world.time`
- All lists are instantiated in New() to prevent an init() call. (minor I know, but a good habit to build)
- All attack_by() calls now return 0 or 1.
- Removed the "Set signal" manual-configuration option from many devices that didn't make use of default/replacement signals.
- User being within range is checked after user-input.
- Devices must now be within a range of 14 of each other to be connected. Fixes an exploit of connecting devices far far away.
- Devices always check for being anchored (level == 2) when receiving input (prevents odd case of a non vended item connecting to it and signaling it)
- Fixed an issue where the Wifi Comp could send out two MechComp signals if NetID & message-type filtering were both off.
- Changed/fixed in the SelectionComp:
  - Fixed an issue where it wouldn't fire in random mode if `current_index` was invalid.
  - Added a bunch of checks to try to keep `current_index` in range more often.
  - Removed a lot of repeated code in favor of calling functions
- ToggleComp can now receive signals without firing.
- TeleComp only fires an outgoing signal when it's `ready`. The signal is the count of atoms teleported. The signal is fired by both the sender and receiver.

Expected to have a merge conflict with #1257
Blocking #1416

## Why's this needed? 
There's a cornucopia of devices that can react to/from MechComp, and this is currently facilitated through /atom/var/mechanics. Not every atom on the map needs this variable, or should need to clear it when disposing. Plus, MechComp behavior seems like a good candidate for Componentization.

Additionally, some of the above are bugs/exploits (in my opinion) and should be patched.

I tested these changes fairly extensively. Not vend-every-device extensive, but I did go through *at least* every device where I changed the flow/logic. In total I think I tested 1/3rd of all the products in the MechComp vendor, as well as the firealarms, light switches, and doors.
## Changelog

```
(u)MarkNstein:
(*)Major refactor of MechComp code. See Minor Changes for details. Please report any bugs!
(+)A MechComp device can not have multiple connections to another device (only one connection per any two devices).
(+)A connection cannot be longer than 14 tiles (one screen height).
(+)TeleComp now sends a signal of how many objects were teleported. Both the sender and receiver pad transmit this number.
(+)SelectionComp is more robust, and shouldn't fail to send a signal as often.
(+)ToggleComp can now receive inputs without firing a signal.
```

### CODE REVIEW

**Comment by system86** [comment]
goddamnit the moment i finish my fancy mechanics thing someone HAS to come sweep the rug out from under it

**Comment by MarkNais** [comment]
@system86 Finish and/or PR your thing. I've already made parity with GlassOfMilks filesending changes during all of this.

**Comment by MarkNais** [comment]
...It is ready...

File: _std/component_defines.dm [hunk 1]
```diff
@@ -82,11 +82,58 @@
...
 #define COMSIG_PROJ_COLLIDE "proj_collide_atom"
 
+// MechComp signals - Content signals - Use these in you MechComp compatible devices
+/// Add an input chanel for a device to send into
+#define COMSIG_MECHCOMP_ADD_INPUT "mechcomp_add_input"
+/// Connect two mechcomp devices together
+#define COMSIG_MECHCOMP_ADD_CONFIG "mechcomp_add_config"
+/// Connect two mechcomp devices together
+#define COMSIG_MECHCOMP_ALLOW_MANUAL_SIGNAL "mechcomp_allow_manual_sigset"
+/// Remove all connected devices     //MarkNstein needs attention
```
Comment (ZeWaka) [inline _std/component_defines.dm:104]: X

File: _std/component_defines.dm [hunk 2]
```diff
@@ -82,11 +82,58 @@
...
+#define COMSIG_MECHCOMP_RM_ALL_CONNECTIONS "mechcomp_remove_all_connections" 
+/// Passing the signal of a message to all connected mechcomp devices for handling (message will be instatiated by the component)
+#define COMSIG_MECHCOMP_TRANSMIT_SIGNAL "mechcomp_transmit_signal"
+/// Passing a message to all connected mechcomp devices for handling
+#define COMSIG_MECHCOMP_TRANSMIT_MSG "mechcomp_transmit_message"
+/// Passing the stored message to all connected mechcomp devices for handling
+#define COMSIG_MECHCOMP_TRANSMIT_DEFAULT_MSG "mechcomp_transmit_default_message"
+
+// MechComp signals - Internal signals - Do not use these
```
Comment (ZeWaka) [inline _std/component_defines.dm:113]: Prefix with a _ ?

File: code/datums/components/mechComp_signals.dm [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+
+/datum/component/mechanics_holder
+	var/list/connected_outgoing
+	var/list/connected_incoming
+	var/list/inputs
+	var/list/configs
+
+	var/defaultSignal = "1"
+
+/datum/component/mechanics_holder/Initialize(can_manualy_set_signal = 0)
```
Comment (ZeWaka) [inline code/datums/components/mechComp_signals.dm:76]: Manually

File: _std/component_defines.dm (hunk 1 above)
Comment (MarkNais) [reply _std/component_defines.dm:104]: Ack, forgot to to "Find in Folder" the **whole** project. Thanks

File: _std/component_defines.dm (hunk 2 above)
Comment (MarkNais) [reply _std/component_defines.dm:113]: Good idea, done.

File: code/datums/components/mechComp_signals.dm (hunk 3 above)
Comment (MarkNais) [reply code/datums/components/mechComp_signals.dm:76]: Thanks

File: code/datums/components/mechComp_signals.dm [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+//Called when a component is dragged onto another one.
+/datum/component/mechanics_holder/proc/dropConnect(var/comsig_target, atom/A, mob/user)
+	if(!A || A == parent || user.stat || !isliving(user) || (SEND_SIGNAL(A,_COMSIG_MECHCOMP_COMPATIBLE) != 1))  //ZeWaka: Fix for null.mechanics
+		return
+
+	if (!user.find_tool_in_hand(TOOL_PULSING))
+		boutput(user, "<span class='alert'>[MECHFAILSTRING]</span>")
+		return
+	
+	if(get_dist(parent, A) > 14)
```
Comment (ZeWaka) [inline code/datums/components/mechComp_signals.dm:220]:
```suggestion
	if(get_dist(parent, A) > SQUARE_TILE_WIDTH)
```

**Comment by MarkNais** [comment]
Of note, the latest pull-from-upstream fixed a conflict with PR#1483.

**Pull request merged**